### ISSUE TITLE
CQL: Unable to maintain performance

### ISSUE THREAD

**Issue by modanesh** [corl-team/CORL#16]
I have trained a CQL agent on HalfCheetah with a good performance, getting to `5000` per episode. I modified the code to save the best model during training not every mode after evaluation:
```python
            if config.checkpoints_path:
                torch.save(
                    trainer.state_dict(),
                    os.path.join(config.checkpoints_path, f"checkpoint_{t}.pt"),
                )
```
to:
```python
            if config.checkpoints_path and eval_score > best_eval_score:
                best_eval_score = eval_score
                torch.save(
                    trainer.state_dict(),
                    os.path.join(config.checkpoints_path, f"best.pt"),
                )
```

The issue happens when I load this model and run it over a newly created environment. In that case, the performance drops to `-700` per episode which is basically the performance of a random agent. Here is the code to test the agent:

```python
@torch.no_grad()
def test_actor(
    config: argparse
) -> np.ndarray:
    env = gym.make(config.env)

    state_dim = env.observation_space.shape[0]
    action_dim = env.action_space.shape[0]
    max_action = float(env.action_space.high[0])

    # Set seeds
    seed = config.seed
    set_seed(seed, env)

    critic_1 = FullyConnectedQFunction(
        state_dim,
        action_dim,
        config.orthogonal_init,
        config.q_n_hidden_layers,
    ).to(config.device)
    critic_2 = FullyConnectedQFunction(state_dim, action_dim, config.orthogonal_init).to(
        config.device
    )
    critic_1_optimizer = torch.optim.Adam(list(critic_1.parameters()), config.qf_lr)
    critic_2_optimizer = torch.optim.Adam(list(critic_2.parameters()), config.qf_lr)

    actor = TanhGaussianPolicy(
        state_dim,
        action_dim,
        max_action,
        log_std_multiplier=config.policy_log_std_multiplier,
        orthogonal_init=config.orthogonal_init,
    ).to(config.device)
    actor_optimizer = torch.optim.Adam(actor.parameters(), config.policy_lr)

    kwargs = {
        "critic_1": critic_1,
        "critic_2": critic_2,
        "critic_1_optimizer": critic_1_optimizer,
        "critic_2_optimizer": critic_2_optimizer,
        "actor": actor,
        "actor_optimizer": actor_optimizer,
        "discount": config.discount,
        "soft_target_update_rate": config.soft_target_update_rate,
        "device": config.device,
        # CQL
        "target_entropy": -np.prod(env.action_space.shape).item(),
        "alpha_multiplier": config.alpha_multiplier,
        "use_automatic_entropy_tuning": config.use_automatic_entropy_tuning,
        "backup_entropy": config.backup_entropy,
        "policy_lr": config.policy_lr,
        "qf_lr": config.qf_lr,
        "bc_steps": config.bc_steps,
        "target_update_period": config.target_update_period,
        "cql_n_actions": config.cql_n_actions,
        "cql_importance_sample": config.cql_importance_sample,
        "cql_lagrange": config.cql_lagrange,
        "cql_target_action_gap": config.cql_target_action_gap,
        "cql_temp": config.cql_temp,
        "cql_alpha": config.cql_alpha,
        "cql_max_target_backup": config.cql_max_target_backup,
        "cql_clip_diff_min": config.cql_clip_diff_min,
        "cql_clip_diff_max": config.cql_clip_diff_max,
    }

    print("---------------------------------------")
    print(f"TESTING CQL, Env: {config.env}, Seed: {seed}")
    print("---------------------------------------")

    # Initialize actor
    trainer = ContinuousCQL(**kwargs)

    model_path = os.path.join(config.checkpoints_path, f"best.pt")
    trainer.load_state_dict(torch.load(model_path, map_location=config.device))
    actor = trainer.actor

    actor.eval()
    episode_rewards = []
    for _ in range(config.test_episodes):
        state, done = env.reset(), False
        episode_reward = 0.0
        while not done:
            action = actor.act(state, config.device)
            state, reward, done, _ = env.step(action)
            episode_reward += reward
        episode_rewards.append(episode_reward)

    print(f"TEST Mean reward: {np.mean(episode_rewards)}")
```

**Comment by modanesh** [comment]
BTW, I did the same with EDAC and it works fine, meaning I'm able to get the same performance after loading a saved actor.

**Comment by Howuhh** [comment]
Hi @modanesh , thanks for noticing, we'll try to figure out what's wrong. Perhaps you could share the checkponit of the model?

**Comment by modanesh** [comment]
Sure. Here it is:
https://drive.google.com/file/d/1T9S68zvgp44uZfjJijaT-kDoHyNqDYz_/view?usp=sharing

**Comment by Howuhh** [comment]
Thanks! Exact config would also help a lot

**Comment by Howuhh** [comment]
I think the issue may be that during training the states were normalized and during testing they were not. Notice this line here:
https://github.com/corl-team/CORL/blob/eeeeef2671235fcce3a3eb255576a207d78840e6/algorithms/offline/cql.py#L855

In our CQL configs we almost always normalize in contrast to the EDAC

**Comment by modanesh** [comment]
Thanks for looking into this. I missed the state normalization part. It's now working fine.
